rust/private/rust_analyzer.bzl

```diff
@@ -0,0 +1,205 @@
+        crate["root_module"] = info.crate.root.path
+        crate_root = info.crate.root.dirname + "/../"
+
+    if info.build_info != None:
+        crate["env"].update({"OUT_DIR": _exec_root_tmpl + info.build_info.out_dir.path})
+        crate["source"] = {
+            # We have to tell rust-analyzer about our out_dir since it's not under the crate root.
+            "include_dirs": [crate_root, _exec_root_tmpl + info.build_info.out_dir.path],
+            "exclude_dirs": [],
+        }
+    crate["env"].update(info.env)
+
+    deps = [
+        {"name": d.crate.name, "crate": crate_mapping[_crate_id(d.crate)]}
+        for d in info.deps
+    ]
+    crate["deps"] = deps
+    crate["cfg"] = info.cfgs
+    crate["target"] = find_toolchain(ctx).target_triple
+    if info.proc_macro_dylib_path != None:
+        crate["proc_macro_dylib_path"] = _exec_root_tmpl + info.proc_macro_dylib_path
+    return crate
+
+# This implementation is incomplete because in order to get rustc env vars we
+# would need to actually execute the build graph and gather the output of
+# cargo_build_script rules. This would require a genrule to actually construct
+# the JSON, rather than being able to build it completly in starlark.
+# TODO(djmarcin): Run the cargo_build_scripts to gather env vars correctly.
+def _rust_project_impl(ctx):
+    rust_toolchain = find_toolchain(ctx)
+    sysroot_src = _exec_root_tmpl + rust_toolchain.rust_lib.label.workspace_root + "/lib/rustlib/src/library"
+
+    # Gather all crates and their dependencies into an array.
+    # Dependencies are referenced by index, so leaves should come first.
+    crates = []
+    crate_mapping = dict()
+    idx = 0
+    for target in ctx.attr.targets:
+        if RustAnalyzerInfo not in target:
+            continue
+
+        # Add this crate's transitive deps to the crate mapping and output.
+        for dep_info in target[RustAnalyzerInfo].transitive_deps.to_list():
+            crate_id = _crate_id(dep_info.crate)
+            if crate_id not in crate_mapping:
+                crate_mapping[crate_id] = idx
+                idx += 1
+                crates.append(create_crate(ctx, dep_info, crate_mapping))
+
+        # Add this crate to the crate mapping and output.
+        crate_id = _crate_id(target[RustAnalyzerInfo].crate)
+        if crate_id not in crate_mapping:
+            crate_mapping[crate_id] = idx
+            idx += 1
+            crates.append(create_crate(ctx, target[RustAnalyzerInfo], crate_mapping))
+
+    # TODO(djmarcin): Use json module once bazel 4.0 is released.
+    ctx.actions.write(output = ctx.outputs.filename, content = struct(
+        sysroot_src = sysroot_src,
+        crates = crates,
+    ).to_json())
+
+rust_analyzer = rule(
+    attrs = {
+        "targets": attr.label_list(
+            aspects = [rust_analyzer_aspect],
+            doc = "List of all targets to be included in the index",
+        ),
+    },
+    outputs = {
+        "filename": "rust-project.json",
```

Comment:
This creates a hard requirement that there's at most one `rust_analyzer` target per package. It would be nice to soften that if we can. Does this _need_ to be called exactly `rust-project.json`? If not, can we mix the name of the target in, like `hello_rust_analyzer-rust-project.json`?

Reply:
It's a much better UX for it to be named rust-project.json because rust-analyzer will automatically detect it without further configuration. More details on the file format here: https://rust-analyzer.github.io/manual.html

Reply:
But that's only needed for the one generated in the top-level WORKSPACE directory, right, not for the templatey one generated in bazel-bin?